File: java/82_RemoveDuplicatesFromSortedListII.java
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode sentinel = new ListNode(0), prev = sentinel, curr = head;
        sentinel.next = head;
        while (curr != null && curr.next != null) {
            if (curr.val == curr.next.val) {
                ListNode tmp = curr.next;
                while (tmp != null && tmp.val == curr.val) tmp = tmp.next;
                
                prev.next = tmp;
                curr = prev.next;
            } else {
                prev = curr;
                curr = curr.next;
            }
        }
        return sentinel.next;
    }
}

File: cpp/82_RemoveDuplicatesFromSortedListII.cpp
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* prev = new ListNode(0);
        prev->next = head;
        ListNode* dummy = prev;
        ListNode* curr = head;
        while (curr && curr->next) {
            if (curr->val == curr->next->val) {
                prev->next = removeDup(curr->next);
                curr = prev->next;
            } else {
                prev = curr;
                curr = curr->next;
            }
        }
        return dummy->next;
    }
    
    ListNode* removeDup(ListNode* head) {
        if (!head) return head;
        ListNode* curr = head->next;
        while (curr && curr->val == head->val) 
            curr = curr->next;
        return curr;
    }
};
